Nirmata provider implementation

augustintsang opened the pull request.

droot committed:
agent: ignore empty query (#485)

droot committed:
docs: update README with latest benchmarks and new sections (#489)

noahlwest committed:
don't return error on failed initialize (#492)

iomarcovalente committed:
Fix Var reference for UI configuration for ui Type (#493)

augustintsang committed:
Merge upstream main and resolve conflicts

- Keep AWS Bedrock 100% benchmark results
- Maintain full Initialize() implementation for Bedrock
- Integrate upstream changes for other providers

augustintsang committed:
fix: resolve tool error handling infinite retry loops

- Fix README.md benchmark table ordering (gemini first, remove bold)
- Implement dynamic status detection in Bedrock integration
  - Check Result.error and Result.status fields for failure indicators
  - Set ToolResultStatusError when errors detected
- Fix MCP client to return structured error data instead of Go errors
  - Preserve error content in JSON format for proper LLM processing
  - Prevents conversion of business logic failures to execution errors

This resolves the issue where all tool results were hardcoded as success,
causing LLMs to retry failed operations indefinitely until max iterations.

augustintsang committed:
Merge remote-tracking branch 'nirmata/nirmata' into bedrock-streaming-minimal

augustintsang committed:
Add Nirmata provider implementation for gollm

Implements HTTP-based Nirmata provider following bedrock.go patterns:
- JWT authentication via NIRMATA_API_KEY header
- Environment variable configuration (NIRMATA_ENDPOINT, NIRMATA_MODEL)
- Simple JSON message format with system prompt in message array
- Both streaming and non-streaming response support
- Client-side conversation history management
- Error handling with proper APIError responses

augustintsang committed:
Clean up and improve Nirmata provider consistency

- Remove unnecessary comments and verbose documentation
- Update default model to match bedrock.go exactly
- Fix ListModels to return same models as bedrock.go
- Remove redundant doRequest method
- Make API key optional for testing
- Improve consistency with bedrock.go patterns

All functionality tested and working with cleaner, more maintainable code.

augustintsang committed:
Fix formatting: add missing newline at end of file

augustintsang committed:
Fix formatting: align multi-line conditionals and remove trailing whitespace

- Fix multi-line conditional alignment in gollm/bedrock.go
- Remove trailing whitespace in pkg/mcp/client.go

augustintsang committed:
Fix formatting: remove trailing whitespace

JimBugwadia commented on gollm/nirmata.go:
```diff
@@ -0,0 +1,542 @@
+
+	baseURLStr := os.Getenv("NIRMATA_ENDPOINT")
+	if baseURLStr == "" {
+		baseURLStr = os.Getenv("NIRMATA_BASE_URL")
```
Why do we need 2 variables for the address?

Lets standardize on one, NIRMATA_ENDPOINT

JimBugwadia commented on gollm/nirmata.go:
```diff
@@ -0,0 +1,542 @@
+var _ Client = &NirmataClient{}
+
+func NewNirmataClient(ctx context.Context, opts ClientOptions) (*NirmataClient, error) {
+	apiKey := os.Getenv("NIRMATA_API_KEY")
```
This will be a JWT. Should we call `NIRMATA_JWT` to be clear?

JimBugwadia commented on gollm/nirmata.go:
```diff
@@ -0,0 +1,542 @@
+	Metadata any    `json:"metadata,omitempty"`
+}
+
+func (c *nirmataChat) Initialize(history []*api.Message) error {
```
is there a way to reuse the Bedrock types and implementation?

augustintsang committed:
Update Nirmata provider environment variables per PR feedback

- Consolidate to single NIRMATA_ENDPOINT environment variable
- Rename NIRMATA_API_KEY to NIRMATA_JWT for clarity
